### PR TITLE
[auth] implement auth context

### PR DESCRIPTION
**Pull request by foloye**
### Description
[//]: # "Required - Describe what's new in this PR in a few lines. A description and bullet points for specifics will suffice."

Auth context (new branch)
Added auth context on to entire app. A user's information can now be accessed in every file of the app without having to query the database.

## How to review
[//]: # 'Required - Describe the order in which to review files and what to expect when testing locally. Is there anything specifically you want feedback on? Should this be reviewed commit by commit, or all at once? What are some user flows to test? What are some edge cases to look out for?'

This can reviewed all at once. In order to test out auth context in pages that require user information, i.e. buttons on profile tab, sign in and sign up pages and test out to see if the app is working as it was previously before auth context was implemented. An edge case to possibly test is calling useSession in a place that a user's information is not truly needed, to make sure that it functions throughout the entire app.

## Next steps
[//]: # "Optional - What's NOT in this PR, doesn't work yet, and/or still needs to be done. Note any temporary fixes in this PR that should be cleaned up later."

DEEP LINKING
After auth context is merged in, next steps are to utilize it so that the app can be more efficient when it comes to calling queries (i.e. making sure that we aren't updating the database when a person edits their full name but it stayed the exact same).

### Online sources
[//]: # 'Optional - copy links to any tutorials or documentation that was useful to you when working on this PR'

### Related PRs
[//]: # "Optional - related PRs you're waiting on/ PRs that will conflict, etc; if this is a refactor, feel free to add PRs that previously modified this code"

CC: @ronniebeggs

### CODE REVIEW

File: src/context/AuthContext.tsx [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+  session: Session | null;
+  user: User | null;
+  isLoading: boolean;
+  signIn: (newSession: Session | null) => void;
+  signUp: (
+    email: string,
+    password: string,
+    options: object,
+  ) => Promise<AuthResponse>;
+  signInWithEmail: (email: string, password: string) => Promise<AuthResponse>;
```
Comment (ronniebeggs) [inline src/context/AuthContext.tsx:37]: Out of curiosity, if the auth context only wraps the `(BottomTabNavigation)` folder, how can the login flow access the `signInWithEmail()` function, for example?

File: src/context/AuthContext.tsx [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    supabase.auth.onAuthStateChange((_event, newSession) => {
+      setSession(newSession);
+    });
+  }, []);
+
+  const signIn = (newSession: Session | null) => {
+    setSession(newSession);
+  };
+
+  const signInWithEmail = async (email: string, password: string) => {
```
Comment (ronniebeggs) [inline src/context/AuthContext.tsx:91]: If we're moving these functions to the auth context, should we delete the duplicate functionality from the `queries/auth.ts` file?

File: src/context/AuthContext.tsx [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+  useContext,
+  useEffect,
+  useMemo,
+  useState,
+} from 'react';
+
+import supabaseAdmin from '../supabase/createAdminClient';
+import supabase from '../supabase/createClient';
+
+export interface AuthState {
```
Comment (ronniebeggs) [inline src/context/AuthContext.tsx:27]: Can we add a few comments/docstrings so other devs know how this works. Our current flow requires that we simply import queries from the `/queries` folder, so we should make this change clear. They don't have to be long. Perhaps a few sentences explaining what the `AuthState` interface represents, for example.

**Review by ronniebeggs — CHANGES_REQUESTED** [review]
Overall it looks good! I had a few minor questions that I'll ask about during work session, but once those are addressed, it should be ready to merge!

File: src/app/_layout.tsx [hunk 4]
```diff
@@ -1,10 +1,15 @@
...
+import { AuthContextProvider } from '../context/AuthContext';
+
 export default function StackLayout() {
   return (
-    <Stack screenOptions={{ headerShown: false }}>
-      <Stack.Screen name="(BottomTabNavigation)" options={{}} />
-    </Stack>
+    <AuthContextProvider>
+      <Stack screenOptions={{ headerShown: false }}>
+        <Stack.Screen name="(Authentication)" options={{}} />
```
Comment (ronniebeggs) [inline src/app/_layout.tsx:10]: I added the `(Authentication)` folder within the AuthContext so it can also utilize the auth functions.

File: src/app/(BottomTabNavigation)/Profile/DeleteAccount/index.tsx [hunk 5]
```diff
@@ -6,16 +6,14 @@ import styles from './styles';
 import WhiteTrash from '../../../../../assets/white-trash.svg';
 import X from '../../../../../assets/x.svg';
 import CasesHeader from '../../../../Components/CasesHeader/CasesHeader';
-import {
-  deleteCurrentUser,
-  getCurrentUserInfo,
-  signOutUser,
-} from '../../../../supabase/queries/auth';
+import { useSession } from '../../../../context/AuthContext';
+import { getCurrentUserInfo } from '../../../../supabase/queries/auth';
```
Comment (ronniebeggs) [inline src/app/(BottomTabNavigation)/Profile/DeleteAccount/index.tsx]: Should this be a function within the `AuthContext`? Ideally we wouldn't have to query this user's information, and it'd already be accessible to them.

**Review by ronniebeggs — CHANGES_REQUESTED** [review]
I did some quick research on auth contexts and this looks like it's coming together well. I guess we can treat this as your PR this cycle and move deep linking to next week? With that, let's make sure each part of the app that uses auth queries now uses the auth context instead! 

Sorry that we didn't merge this earlier. I was focused on merging front facing progress to show our NPO contact and this slipped my mind. It looks great though!

File: src/app/(BottomTabNavigation)/Profile/DeleteAccount/index.tsx (hunk 5 above)
Comment (foloye) [reply src/app/(BottomTabNavigation)/Profile/DeleteAccount/index.tsx]: youre right this should be deleted and not used

**Review by ronniebeggs — APPROVED** [review]
Everything looks good to me! Great job Folasade!
